QA-232: suggest cherry-picks to the latest three release branches

tranchitella opened the pull request.

tranchitella commented:
@lluiscampos any suggestion on how to get rid of the `Host key verification failed` (see the failed job in the CI pipeline)?

tranchitella commented:
If we want to run unit tests, I need the goveralls setup. Before proceeding, though, I'm waiting for your feedbacks @lluiscampos and @oleorhagen

oleorhagen commented on menderqa_pipeline_helpers.go:
```diff
@@ -9,8 +9,21 @@ import (
+
 func getServiceRevisionFromIntegration(repo, baseBranch string, conf *config) (string, error) {
-	c := exec.Command("release_tool.py", "--version-of", repo, "--in-integration-version", baseBranch)
+	c := exec.Command("python3", "release_tool.py", "--version-of", repo, "--in-integration-version", baseBranch)
```
Question:
Why was this needed?

tranchitella replied:
Because it doesn't work if you don't have "." in your PATH (like on my Mac OS).
Unit tests will fail because release_tool.py is not executable directly. It's either "./release_tool.py" or "python3 release_tool.py".

oleorhagen reviewed (approved):
I think it looks good.

Prior to reviewing this I was thinking along the lines, that maybe we should have the ability to modify the predicate it matches on for notifying about cherries needed (like a function which takes the commit message (string), and returns true of false, and is then mapped onto the commit messages in order). However, it's not necessary. 

Also, prior to reviewing I wasn't sure if `Changelogs` are the best measure here. But you guys are probably right (I was thinking along the lines of having something like `fix` or `fixes` in the commit message would trigger this.).

kacf commented on cherrypicks.go:
```diff
@@ -0,0 +1,142 @@
+
+	// count the number of non-none changelog entries
+	baseSHA := *pr.PullRequest.Base.SHA
+	countCmd := exec.Command("sh", "-c", "git log "+baseSHA+"...pr_"+prNumber+" | grep -i \"Changelog: \" | grep -v -i \": none\" | wc -l")
```
Suggest being a bit more explicit, since it's not uncommon to include the word "changelog" in various other contexts for example when doing releases, and to have the word "none" in an actual changelog message:

```
grep -i \"^Changelog:\" | grep -v -i \"^Changelog: *none\"
```

Might be useful to have a test for these sentences too. Examples would be:

* ```
  Fix changelog: This was actually reverted.

  Cancel-changelog: abcdef0123456789
   ```
* ```
  Check for nil.

  Changelog: Fix API returning none.
```

tranchitella replied:
Nice, thanks! Your grep machinery is way better than mine, I applied it.

kacf commented on deb-requirements.txt:
```diff
@@ -0,0 +1 @@
+python3-yaml
```
Not if the dependency on this component is introduced in this commit.

tranchitella replied:
I need it to run unit tests, why do I need to separate it in a different commit?

oleorhagen replied:
I see, I figured it was missing from old.

kacf commented on menderqa_pipeline_helpers.go:
```diff
@@ -9,8 +9,21 @@ import (
 
+func getLatestIntegrationRelease(number int, conf *config) ([]string, error) {
+	cmd := fmt.Sprintf("git for-each-ref --sort=-creatordate --format='%%(refname:short)' 'refs/tags' "+
+		"| sed -E '/(^[0-9]+\\.[0-9]+)\\.[0-9]+$/!d;s//\\1.x/' | uniq | head -n %d | sort -n -r", number)
```
Suggestion: Instead of `sort -n`, use `sort -V`, which is specifically for sorting versions.

tranchitella commented on entfork_sync.go:
```diff
@@ -14,8 +14,9 @@ import (
-
 	"github.com/sirupsen/logrus"
+
+	tgithub "github.com/mendersoftware/integration-test-runner/client/github"
```
because "github" is the import of the GitHub library, I need to use an alias because it is clashing with my github client, which I introduced to be able to mock it in unit tests.

oleorhagen replied:
Yes, but `t` for `temp`? :thinking:

tranchitella replied:
t for `test-runner`, that was my intention :)

tranchitella replied:
let's rename it to client_github

kacf commented on main_test.go:
```diff
@@ -0,0 +1,23 @@
+	}
+	// import the SSH private key
+	cmd = exec.Command("sh", "-c", "[ -n \"$SSH_PRIVATE_KEY\" ] && ! [ -f $HOME/.ssh/id_rsa ] && "+
+		"echo \"$SSH_PRIVATE_KEY\" | tr -d '\r' | tee $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa || true")
```
We should not use `tee` for a private key, suggest using straight redirection so it doesn't show up in the log.

lluiscampos commented on README.md:
```diff
@@ -21,35 +21,29 @@ For all repositories in the organization, a pr_XXX branch will be created in
 GitLab for every pull/XXX PR from GitHub.
 
```
Thanks for updating README

tranchitella commented on cherrypicks.go:
```diff
@@ -0,0 +1,142 @@
+		return fmt.Errorf("%v returned error: %s: %s", gitcmd.Args, out, err.Error())
+	}
+
+	changelogs, _ := strconv.Atoi(strings.TrimSpace(string(out)))
```
TBH, I like strconv.Atoi. Can you elaborate more on the advantages of binary/encoding?

oleorhagen replied:
You just don't have to go via strings, but as mentioned: `feel free to choose`. And it seems you have :)

lluiscampos commented on menderqa_pipeline_helpers.go:
```diff
@@ -9,8 +9,21 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+func getLatestIntegrationRelease(number int, conf *config) ([]string, error) {
```
This function does not belong here: `menderqa_pipeline_helpers.go` are meant to be used only from `menderqa_pipeline.go`. Maybe you want to put it in `util.go`, or directly in `cherrypicks.go` (it is not used anywhere else).

tranchitella replied:
moved

lluiscampos commented on main.go:
```diff
@@ -264,14 +251,19 @@ func processGitHubWebhook(ctx *gin.Context, payload []byte, githubClient *github
 			log.Errorf("Failed to sync the OS and Enterprise repos: %s", err.Error())
 		}
+
+		// If the pr was merged, suggest cherry-picks
+		if err := suggestCherryPicks(log, pr, githubClient, conf); err != nil {
+			log.Errorf("Failed to suggest cherry picks for the pr %v. Error: %v", pr, err)
+		}
```
This has to be done above line 240, or else won't be processed for PRs from contributors.

tranchitella replied:
right, done

lluiscampos commented on main_test.go:
```diff
@@ -0,0 +1,23 @@
+		panic(err)
+	}
+	// import the SSH private key
+	cmd = exec.Command("sh", "-c", "[ -n \"$SSH_PRIVATE_KEY\" ] && ! [ -f $HOME/.ssh/id_rsa ] && "+
+		"echo \"$SSH_PRIVATE_KEY\" | tr -d '\r' >> $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa || true")
+	err = cmd.Run()
```
I don't understand what is the intention here. We are not going to set an SSH_PRIVATE_KEY in CI variables, so is only meant for local use? Can the tests run without the SSH key at all?

tranchitella replied:
Without the SSH key, the unit tests cannot clone the repos, thus unit tests won't run in CI. Am I wrong?

lluiscampos replied:
Exactly, but we cannot have an *unprotected* SSH key on CI variables. I saw we have it right now and I assumed you added it while developing, but we definitely have to remove before merge, so the tests won't be able to run.

tranchitella replied:
ok, which options do I have, then? maybe we can discuss it on slack.

tranchitella replied:
reworked to work with https links when running unit tests

lluiscampos reviewed (changes requested):
Having already @kacf and @oleorhagen eyes on the code, I focused more on high level things.

See my comments below; the only one "requesting for changes" is the one in `main.go`, as I believe we need to move this check earlier in the process.

Thank you very much for introducing unit tests. This is going to help everyone working on the tool :heart:

tranchitella commented:
@lluiscampos I removed the SSH private key variable in GitLab and implemented here (https://github.com/mendersoftware/integration-test-runner/pull/76/commits/be199cb754733167d53125f30157375411f9e34c) the possibility to use HTTPS urls in unit tests.
Can you please review the PR again?

lluiscampos commented on util_test.go:
```diff
@@ -0,0 +1,15 @@
+	"gotest.tools/assert"
+)
+
+func TestGetRemoteURLGitHub(t *testing.T) {
+	url := getRemoteURLGitHub(GitProtocolSSH, "mendersoftware", "workflows")
+	assert.Equal(t, "git@github.com:/mendersoftware/workflows.git", url)
+
+	url = getRemoteURLGitHub(GitProtocolHTTP, "mendersoftware", "workflows")
+	assert.Equal(t, "https://github.com/mendersoftware/workflows", url)
+}
```
:+1:

lluiscampos reviewed (approved):
@tranchitella Excellent work. Let's deploy it!
